### ISSUE TITLE
Filter out duplicate POIs

### ISSUE THREAD

**Issue by zerebubuth** [mapzen/vector-datasource#267]
Filtering out duplicate POIs is best done as a step near the end of processing, as we can use earlier filtering steps to discard attributes (zoom appropriate) which aren't important. This means we can merge more POIs, as there's less "different" information on them with fewer properties.

**Comment by nvkelso** [comment]
This is probably why I'm seeing 2 to 3 label_positions for some churches, sports stadiums & etc. The one without area is often stacking on the one with area making debug fun.

Example:

* http://vector.mapzen.com/osm/all/16/10481/25328.topojson?api_key=vector-tiles-HqUVidw

**Comment by nvkelso** [comment]
Food for thought: 

**Lincoln Park Golf Course** is a multi polygon in SF.  We probably want to allow more than one icon here, but only at later zooms (like z15+, or if the mutlipolygon parts are far apart from each other)?

https://www.openstreetmap.org/relation/2142043
#15/37.7828/-122.4939

**Comment by nvkelso** [comment]
Piers in SF: http://mapzen-tiles-prod-us-east-20151007.s3.amazonaws.com/osm/all/19/83894/202606.topojson?api_key=vector-tiles-HqUVidw

**Comment by nvkelso** [comment]
Church: http://mapzen-tiles-prod-us-east-20151007.s3.amazonaws.com/osm/all/20/167692/405409.topojson?api_key=vector-tiles-HqUVidw

**Comment by zerebubuth** [comment]
> Church: http://mapzen-tiles-prod-us-east-20151007.s3.amazonaws.com/osm/all/20/167692/405409.topojson?api_key=vector-tiles-HqUVidw

Looks like this is fixed with the latest code:

![267-saint-pauls-church](https://cloud.githubusercontent.com/assets/271360/10524197/330637b4-7376-11e5-94e3-6cca855dc03a.png)

Does that look okay?

**Comment by zerebubuth** [comment]
> Piers in SF: http://mapzen-tiles-prod-us-east-20151007.s3.amazonaws.com/osm/all/19/83894/202606.topojson?api_key=vector-tiles-HqUVidw

Just checking - what you want is for Pier 1, Pier 3 and Pier 7 to all merge together? Or was it just the building + POI label which should have been fixed by the landuse/building/POI cleanup that just got merged. Let me know - I wouldn't want to go merging things which shouldn't be merged!

![267-piers](https://cloud.githubusercontent.com/assets/271360/10524252/92b7a918-7376-11e5-91b5-5d3c61794ebc.png)

**Comment by zerebubuth** [comment]
> Lincoln Park Golf Course is a multi polygon in SF. We probably want to allow more than one icon here, but only at later zooms (like z15+, or if the mutlipolygon parts are far apart from each other)?

Sadly, these are in separate tiles starting at zoom 13 when they first get displayed... which rules out the way I wanted to implement this :disappointed:

Just to make certain: When we have multiple separate polygons being labelled from the same feature (e.g: this golf course) then we want to label only the largest until such a zoom as the centroids are some (configurable) distance apart?

Would we want to add the areas of the parts not being displayed onto the parts which are displayed? This would potentially decrease their `min_zoom` and might be beneficial for large features composed of many smaller parts.

**Comment by nvkelso** [comment]
> Sadly, these are in separate tiles starting at zoom 13 when they first get displayed

Another reason for moving towards meta tiles? How much work is there left on solving this for the "same tile" case with the other changes you're landing? If the bulk of duplicate POIs (on top of each other) are fixed with that work then maybe we should save this "neighboring tile" & multi polygon case for later meta tile work?

**Comment by zerebubuth** [comment]
> Another reason for moving towards meta tiles?

The way we'd planned to do metatiles wouldn't affect this - it was basically an optimisation only. We wanted to avoid adding extra complexity and code paths for bugs to hide in.

> How much work is there left on solving this for the "same tile" case with the other changes you're landing?

Since we generate labels from the polygons now, then I think there's some hope of tackling this case if we do the post-process step at the right point. Since the polygons overlap both tiles, then I think it might work. If so, then I should have something ready later today.

> If the bulk of duplicate POIs (on top of each other) are fixed with that work then maybe we should save this "neighboring tile" & multi polygon case for later meta tile work?

If the bulk of duplicates on top of each other are fixed - which I think we should be able to check in dev soon - and what I'm playing with now doesn't work, then we might want to consider putting this back on the backlog and coming back to it later. I'm not sure that it's worth coupling this to the metatile work, though, as I think they target different things (i.e: this is a cartography issue, metatiles are a performance issue).

**Comment by nvkelso** [comment]
> Since the polygons overlap both tiles, then I think it might work. If so, then I should have something ready later today.

Looking forward to it :)

**Comment by zerebubuth** [comment]
> Looking forward to it :)

FAIL. Sorry about that.

After #313 and mapzen/TileStache#74, I can get screenshots like this:

## Zoom 15, Lincoln Park Golf Course

Note that there's only one POI for the golf course at this zoom.

![lincoln-park-golf-course-z15](https://cloud.githubusercontent.com/assets/271360/10559355/dc2b6e06-74e6-11e5-9e0b-3b490baa095b.jpg)

## Zoom 16

...and now there's two. The distance is configurable, but I pretty much had to set it to the maximum that makes sense for a single tile to get this behaviour. If we need longer exclusion distances we should think about different (i.e: harder) ways of doing it, or perhaps doing it client-side.

![lincoln-park-golf-course-z16](https://cloud.githubusercontent.com/assets/271360/10559367/0c94c196-74e7-11e5-99b3-ee5db04462da.jpg)

**Comment by nvkelso** [comment]
Looking better!

But I'm still seeing duplicate POIs, I think from **different layers** where one has an area and the other doesn't, but they have the same OSM ids. The proposed implementation is only within a single layer, do we need it to span across layers, or pack less stuff into multiple layers? (Seems like less double packing is the way to go.)

#16/37.7653/-122.4413
![screen shot 2015-10-19 at 17 08 18](https://cloud.githubusercontent.com/assets/853051/10594703/bffb413a-7684-11e5-9299-b89b37e065c8.png)

**Comment by zerebubuth** [comment]
I think the tile with the museum must have just not updated yet - what I see is below:

![randall_museum_z16](https://cloud.githubusercontent.com/assets/271360/10604656/b83f1fa4-771d-11e5-90af-bdf10640bbb4.jpg)

Regarding the duplicate schools; looks like they're coming from an [`amenity=school` polygon](http://www.openstreetmap.org/way/300732288) which we put into the landuse layer and gets a label from there, and a [`building=school` polygon](http://www.openstreetmap.org/relation/3495920) which we put into the buildings layer and gets a label from there.

We have two options here:

1. Noting that this is very similar to the situation around the SFU campus, where buildings are appearing like POIs because their kind ends up being the same as the landuse polygon containing them, you had previously suggested suppressing the kind when it's the same as the `landuse_kind` and turn it into `kind=building` instead. We could do this as a separate post-process filter. (Alternatively, we could suppress the landuse polygon, but I think we'd prefer to keep the "larger" object, which is usually the landuse not the building polygon.)
2. Suppress duplicate POIs (across layers) by extending the post-processing step. This will also run into more corner cases, as we'd effectively need to define an ordering between layers so that we could choose the more important duplicate to keep.

In either case, I think it's worth tracking the follow-up as a separate issue so as not to hold up getting this, or anything merged after this, into prod.

**Comment by nvkelso** [comment]
Cool, let's track followup work with: https://github.com/mapzen/vector-datasource/issues/320

**Comment by nvkelso** [comment]
Working for me.
